resources/migrations/000_migrations.yaml

```diff
@@ -4020,3 +4020,23 @@ databaseChangeLog:
                   remarks: 'True if a XLS of the data should be included for this pulse card'
                   constraints:
                     nullable: false
+  - changeSet:
+      id: 75
+      author: camsaul
+      comment: 'Added 0.28.2'
+      changes:
+        - addColumn:
+            tableName: report_card
+            columns:
+              name: read_permissions
+              type: text
+              remarks: 'Permissions required to view this Card and run its query.'
+        # We want to enforce NOT NULL on read_permissions, but to populate it for existing Cards we'll have to do so in
+        # Clojure-land. Thus we'll give all the existing Cards an entry of `null` which the JSON parser will parse as `nil`,
+        # which is basically the same thing as these columns being `NULL` -- nice! But going forward new Cards won't be able
+        # to be saved with NULL values here
```

Comment:
This is a bummer since "null" will cause the migration to change every row, then we'll change every row again to swap "null" for the real value. I guess the only way to make that go away is to somehow link a db migration with a data migration, i.e. run id 75 with a nullable constraint, run data migration, apply the not null constraint with or without the defaultNull.

Reply:
Yeah. I didn't think about that. I guess another option would be to just forget about adding the  `NOT NULL` constraint entirely. Either way even with 20k so-called "huge" instances of Metabase this shouldn't be too bad. Even with 20k Cards this should be reasonably quick. (It's not as big of an issue as if we were doing something like this to Field, for example.)

Reply:
@senior I'm thinking maybe it just makes more sense to drop the `NOT NULL` constraint since if the column is `NULL` nobody is going to die. Also since `pre-insert` and `pre-update` are taking care of populating it the extra safety check will be a little less necessary. Thoughts?

Reply:
Seems like a good idea. There's a few columns that I think are in this same situation. Theoretically NOT NULL, but for practical reasons, nullable. Might be worth eventually circling back and cleaning them up, but all at once.